### PR TITLE
[FLINK-36361][table] Do not use StringBuilder for EquivalentExprShuttle

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

The idea is to use just set of RelNodes and do not generate toString each time.
Also use the fact that Calcite can compare RelNodes

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable )

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/rules/logical/JoinDependentConditionDerivationRuleTest.xml [hunk 1]
```diff
@@ -98,7 +98,7 @@ LogicalProject(a=[$0], d=[$3])
     <Resource name="optimized rel plan">
       <![CDATA[
 LogicalProject(a=[$0], d=[$3])
-+- LogicalJoin(condition=[AND(OR(AND(=($0, 1), =($1, 1), =($3, 2), =($4, 2)), AND(=($0, 2), =($1, 2), =($3, 1), =($4, 1))), OR(AND(=($0, 1), =($1, 1)), AND(=($0, 2), =($1, 2))), OR(AND(=($3, 2), =($4, 2)), AND(=($3, 1), =($4, 1))))], joinType=[inner])
++- LogicalJoin(condition=[AND(OR(AND(=($0, 1), =($1, 1), =($3, 2), =($4, 2)), AND(=($0, 2), =($1, 2), =($3, 1), =($4, 1))), OR(AND(=(1, $0), =(1, $1)), AND(=(2, $0), =(2, $1))), OR(AND(=(2, $3), =(2, $4)), AND(=(1, $3), =(1, $4))))], joinType=[inner])
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/rules/logical/JoinDependentConditionDerivationRuleTest.xml:101]: here and in others just swap of left and right

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala [hunk 2]
```diff
@@ -439,17 +439,16 @@ object FlinkRexUtil {
...
+    private val equiExprSet = mutable.HashSet[RexNode]()
 
     override def visitCall(call: RexCall): RexNode = {
       call.getOperator match {
         case EQUALS | NOT_EQUALS | GREATER_THAN | LESS_THAN | GREATER_THAN_OR_EQUAL |
             LESS_THAN_OR_EQUAL =>
-          val swapped = swapOperands(call)
-          if (equiExprMap.contains(swapped.toString)) {
-            swapped
+          if (equiExprSet.contains(call)) {
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:448]: For Calcite `>($1, $2)` and `<($2, $1)` are equal, same for others
this fact could be used here.

Ideally we could reuse Calcite's RexNormalize however it will lead to large plan related changes, may be makes sense to do it with separate PR/jira

**Review by jnh5y — APPROVED** [review]
This change makes sense to me.  I think saving on the `toString` calls is an improvement.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala (hunk 2 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:448]: Did you already try the RexNormalize change?

That does sound like a good idea / nice change.  Have you filled a JIRA for it?

**Comment by snuyanzin** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala (hunk 2 above)
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:448]: jira issue https://issues.apache.org/jira/browse/FLINK-36367

I tried
1. it touches more than 100 plans which is still doable
2. I faced one float issue for a couple of plans: depending on how tests are running (just one or for the whole class) the order of operands is different (the operator expression is valid for both).
So need more time to debug and fix it

**Comment by snuyanzin** [comment]
thanks for taking a look

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala (hunk 2 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala:448]: > 2. I faced one float issue for a couple of plans: depending on how tests are running (just one or for the whole class) the order of operands is different (the operator expression is valid for both).
>    So need more time to debug and fix it
Interesting.  Is it a bug in RexNormalize or something else?
